runtime/gc_vlhgc/CopyForwardScheme.cpp

```diff
@@ -4063,21 +4066,66 @@ class MM_CopyForwardSchemeRootClearer : public MM_RootScanner
-			if (NULL == objectPtr) {
+			void *forwardedObject = forwardedHeader.getForwardedObject();
+			bool virtualLargeObjectHeapEnabled = _extensions->indexableObjectModel.isVirtualLargeObjectHeapEnabled();
+			/* If forwardedObject is NULL, free the double mapped region occupied by the data of the indexable object */
+			if (NULL == forwardedObject) {
 				Assert_MM_mustBeClass(_extensions->objectModel.getPreservedClass(&forwardedHeader));
-				env->_copyForwardStats._doubleMappedArrayletsCleared += 1;
+				env->_copyForwardStats._doubleMappedOrVirtualLargeObjectHeapArrayletsCleared += 1;
 				OMRPORT_ACCESS_FROM_OMRVM(_omrVM);
-				omrvmem_release_double_mapped_region(identifier->address, identifier->size, identifier);
+				/* If forwardedObject is NULL and virtualLargeObjectHeapEnabled is true, free the sparse region occupied by the data of the indexable object */
+				if (virtualLargeObjectHeapEnabled) {
+					_extensions->largeObjectVirtualMemory->freeSparseRegionAndUnmapFromHeapObject(_env, dataAddr);
+					_extensions->indexableObjectModel.setDataAddrForContiguous((J9IndexableObject *)objectPtr, NULL);
+				} else {
+					omrvmem_release_double_mapped_region(identifier->address, identifier->size, identifier);
+				}
+			} else if (virtualLargeObjectHeapEnabled && NULL != dataAddr) {
+				/* There might be the case that GC finds a floating arraylet, which was a result of an allocation
+				 * failure (reason why this GC cycle is happening). */
+				if (!_extensions->indexableObjectModel.isAddressWithinHeap(_extensions, dataAddr)) {
+					_extensions->largeObjectVirtualMemory->updateSparseDataEntryAfterObjectHasMoved(dataAddr, forwardedObject);
+				}
 			}
 		}
 	}
 #endif /* J9VM_GC_ENABLE_DOUBLE_MAP */
 
+#if defined(J9VM_ENV_DATA64)
+	virtual void doObjectInVirtualLargeObjectHeap(J9Object *objectPtr) {
+		MM_EnvironmentVLHGC *env = MM_EnvironmentVLHGC::getEnvironment(_env);
+		env->_copyForwardStats._doubleMappedOrVirtualLargeObjectHeapArrayletCandidates += 1;
+
+		if (!_copyForwardScheme->isLiveObject(objectPtr)) {
+			Assert_MM_true(_copyForwardScheme->isObjectInEvacuateMemory(objectPtr));
+			void *dataAddr = _extensions->indexableObjectModel.getDataAddrForContiguous((J9IndexableObject *)objectPtr);
+			MM_ForwardedHeader forwardedHeader(objectPtr, _extensions->compressObjectReferences());
+			objectPtr = forwardedHeader.getForwardedObject();
+			void *forwardedObject = forwardedHeader.getForwardedObject();
+			/* If forwardedObject is NULL, free the sparse region occupied by the data of the indexable object */
+			if (NULL == forwardedObject) {
+				Assert_MM_mustBeClass(_extensions->objectModel.getPreservedClass(&forwardedHeader));
+				env->_copyForwardStats._doubleMappedOrVirtualLargeObjectHeapArrayletsCleared += 1;
+				_extensions->largeObjectVirtualMemory->freeSparseRegionAndUnmapFromHeapObject(_env, dataAddr);
+				_extensions->indexableObjectModel.setDataAddrForContiguous((J9IndexableObject *)objectPtr, NULL);
```

Comment:
I'd rather leave this pointer as-is, rather than null it. It may be useful in some debugging. @dmitripivkine, agree?
In general case, we don't null any references of any object that died.